### PR TITLE
TTS source improvements 

### LINKED ISSUE #51: TTS source improvements
#### Summary

References https://github.com/TheThingsIndustries/lorawan-stack-support/issues/854#issuecomment-1260510806

We should add extra options to the TTS (`ttnv3`) source.

#### Why do we need this?

These options would allow easier migration.

#### What is already there? What do you see now?

Non dry-run runs of the tool disable the end device downlinks and export the end device.

#### What is missing? What do you want to see?

1. A `--no-session` option that would enable users to export the device without the session added. 
2. A `--delete-source-device` option that would delete the end device in the 4 source servers (JS, NS, AS, IS. The order matters to a certain extent. The IS should always be the last one).

For the above, `--dry-run` should cause these options to not be executed.

#### Environment

`v0.9.0-rc1`.

#### How do you propose to implement this?

1. We do a `SetFields` from `nil` with a field mask of `"session", "pending_session", "mac_state", "pending_mac_state"`, in the same fashion that we delete the addresses.
2. It is similar to device updates, but you call it with `Delete` instead of `Set`.

#### How do you propose to test this?

Use the options and check that:
1. There is no `session`, `pending_session`,`mac_state` or `pending_mac_state` in the returned end device.
2. The device is delete after you run the command. You can check that everything went fine by attempting to re-register it.

#### Can you do this yourself and submit a Pull Request?

Can review.

### PR DESCRIPTION
**Pull request by happyRip**
#### Summary

Closes #51 

#### Changes

- Add requested flags:
  - `--ttnv3.no-session`
  - `--ttnv3.delete-source-device`

#### Testing

Tested both flags locally.

##### Regressions

n/a

#### Notes for Reviewers

nil

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

### CODE REVIEW

**Comment by jpmeijers** [comment]
In the process of trying to work around https://github.com/TheThingsNetwork/lorawan-stack/issues/5926 I used this branch to export a json file of all devices belonging to an application, without their sessions. Using this command:

```
$ go run ./cmd/ttn-lw-migrate application rbs304 --source ttnv3 --dry-run --verbose --ttnv3.no-session > rbs304.json
```

It provided me a json file that imported successfully into TTS Cloud v3.22.2, without setting the sessions, and also not triggering the `store error`.

File: pkg/source/ttnv3/source.go [hunk 1]
```diff
@@ -296,3 +309,34 @@ func (s Source) setEndDevice(device *ttnpb.EndDevice, isPaths, nsPaths, asPaths,
 	}
 	return &res, nil
 }
+
+func (s Source) deleteEndDevice(ids *ttnpb.EndDeviceIdentifiers) error {
+	if address := s.config.identityServerGRPCAddress; address != "" {
```
Comment (adriansmares) [inline pkg/source/ttnv3/source.go:314]: The order here should be in reverse (AS, NS, JS, IS) , and you are not deleting the entity from the JS.

File: pkg/source/ttnv3/source.go [hunk 2]
```diff
@@ -74,13 +74,26 @@ func (s Source) ExportDevice(devID string) (*ttnpb.EndDevice, error) {
...
 
-	if !s.config.dryRun {
+	if s.config.noSession {
+		if err := clearDeviceSession(dev); err != nil {
+			return nil, err
+		}
+	}
+	switch {
+	case s.config.deleteSourceDevice:
+		ids := dev.GetIds()
```
[the remark this answers is not in the document]
Comment (happyRip) [reply pkg/source/ttnv3/source.go]: Fixed

File: pkg/source/ttnv3/source.go (hunk 1 above)
Comment (happyRip) [reply pkg/source/ttnv3/source.go]: Reordered and added missing JS

**Review by adriansmares — APPROVED** [review]

**Comment by jpmeijers** [comment]
When `no-sessions` is used, the last activity of the old network is still copied over to the new network. This makes it more difficult to know which devices still need to join the new network.

Maybe it would be better if we also remove:
* `last_dev_status_received_at`
* `activated_at`
* `last_seen_at`

**Pull request merged**